### ISSUE TITLE
Depth map overfitting on one datapoint fails when training from scratch

### ISSUE THREAD

**Issue by m43** [facebookresearch/vggt#221]
I have trouble overfitting VGGT on one datapoint when training from scratch. I'd expect the model to perfectly overfit the datapoint, however the predicted depth map is always blurry around object edges. To isolate the issue, I'm only regressing the depth map for a single view with a simple L2 regression loss for one datapoint. I remove all other losses: for camera, tracking, point map, and the depth confidence and gradient losses. I've tried with different datapoints, LRs, LR schedules, L1/L2 loss, number of training steps, full precision, and more, but always get a blurry depth map. Surprisingly, finetuning from pretrained weights overfits perfectly. Also surprisingly, short-circuiting the DPT Head (by replacing it with a shallow MLP) also overfits perfectly. Short-circuiting the aggregator (so `rgbs-->mlp-->depth_head-->depth_map`) will not overfit. Do you know more about this, did you face similar issues when overfitting, is this known DPT behavior? Below are examples of depth maps visualized as points (using GT camera parameters) and turbo-cmap images:

<table>
  <tr>
    <th>GT</th>
    <th>Trained from scratch</th>
    <th>DPTHead short-circuited</th>
    <th>Finetuned from checkpoint </th>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/519c6b42-8e24-488f-bc1b-5b4f135685f0" width="250"/></td>
    <td><img src="https://github.com/user-attachments/assets/a668d1d5-eeec-4123-baa6-06c1762c6be0" width="250"/></td>
    <td><img src="https://github.com/user-attachments/assets/40f4526b-2d43-4d6c-9e02-ccd39ad85d8c" width="250"/></td>
    <td><img src="https://github.com/user-attachments/assets/986e79ac-1fb4-48fd-879b-5b5b418e0046" width="250"/></td>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/65845ef3-1fa5-4bbd-abfd-09d9fc8d31ea" width="250"/></td>
    <td><img src="https://github.com/user-attachments/assets/636d2916-b6c9-4ecd-96ab-8ffdaff4a160" width="250"/></td>
    <td><img src="https://github.com/user-attachments/assets/ba2071fe-8b28-4d10-b0b9-adea851409f8" width="250"/></td>
    <td><img src="https://github.com/user-attachments/assets/39d062d1-38db-49e8-b1dd-40c0523b8f3b" width="250"/></td>
  </tr>
</table>

**Comment by jytime** [comment]
Hi Frano,

Great digging! I’m currently cleaning the training code, so I might not have enough time to debug this myself. From what you shared, it looks related to a known issue with DPT: strong training instability, especially under small effective batch sizes.

To verify this, would you mind trying, still overfitting, while using a very large batch size?

**Comment by m43** [comment]
Yes, I'll look into that!

**Comment by m43** [comment]
It's unfortunately the same pattern, for example this is after 10K iterations with an effective batch size of 192:
<table>
  <tr>
    <th>GT</th>
    <th>BS=192</th>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/03ed18a0-c16c-411e-8452-b2971d943c0c" width="500"/></td>
    <td><img src="https://github.com/user-attachments/assets/b7255957-d744-42e5-9c71-8fa9d5609aef" width="500"/></td>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/028fcb47-a89c-4235-8e97-84228271d4f0" width="500"/></td>
    <td><img src="https://github.com/user-attachments/assets/2843165f-3da9-4ed7-9750-d25509db43a7" width="500"/></td>
  </tr>
</table>

However, I did notice that training for much much longer, for example 20K iterations instead of 500 iterations with batch size of 1, does overfit better, although still not perfectly well as I would expect or hope it would since some points are still predicted in the air around edges:
<table>
  <tr>
    <th>GT</th>
    <th>500 steps</th>
    <th>5K steps</th>
    <th>20K steps</th>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/013b8623-3b4a-47b2-b76b-c3a5e2f9e9fb" width="250"/></td>
    <td><img src="https://github.com/user-attachments/assets/94b4d8e6-900a-4330-b268-ff2d85da8ec4" width="250"/></td>
    <td><img src="https://github.com/user-attachments/assets/49cbbc40-13b3-4162-a4d5-0f08275da9d0" width="250"/></td>
    <td><img src="https://github.com/user-attachments/assets/ebdd9bbe-3106-4c41-b2ca-9b8a4cffa46b" width="250"/></td>
  </tr>
</table>

I'll think about whether I look more into DPT's behavior or whether I move on from here since the overfitting is reasonable, but let me know if you have more ideas or things to try
